File: DB/10월 4주차/게임맵최단거리.js
function solution(maps) {
  const n = maps.length;
  const m = maps[0].length;

  // 네 방향 이동 (상, 하, 좌, 우)
  const directions = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];

  // BFS 초기 설정
  const queue = [[0, 0, 1]];
  const visited = Array.from(Array(n), () => Array(m).fill(false));
  visited[0][0] = true;

  while (queue.length > 0) {
    const [x, y, distance] = queue.shift();

    // 목표 위치에 도달한 경우
    if (x === n - 1 && y === m - 1) {
      return distance;
    }

    // 네 방향으로 이동
    for (const [dx, dy] of directions) {
      const nx = x + dx;
      const ny = y + dy;

      // 맵의 범위 안이고, 벽이 아니며, 방문하지 않은 경우 이동
      if (
        nx >= 0 &&
        ny >= 0 &&
        nx < n &&
        ny < m &&
        maps[nx][ny] === 1 &&
        !visited[nx][ny]
      ) {
        visited[nx][ny] = true;
        queue.push([nx, ny, distance + 1]);
      }
    }
  }

  // 목표 위치에 도달하지 못한 경우
  return -1;
}

File: JY/10월 5주차/게임맵최단거리.java
import java.util.*;

class Solution {
    public static int[][] grid;
    public static int[] dx = {0, 0, 1, -1};
    public static int[] dy = {1, -1, 0, 0};
    public static class Point{
        int x;
        int y;
        
        public Point(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
    
    public static boolean is_possible(int nx, int ny){
        if(nx < 0 || nx >= grid[0].length)
            return false;
        if(ny < 0 || ny >= grid.length)
            return false;
        return true;
    }
    
    public static void print(int[][] arr){
        for(int i=0; i<arr.length; i++){
            for(int j=0; j<arr[i].length; j++){
                System.out.print(arr[i][j] + " ");
            }
            System.out.println();
        }
    }
    
    public static int bfs(Point start){
        Queue<Point>q = new LinkedList<>();
        q.add(start);
        
        int[][] dis = new int[grid.length][grid[0].length];
        dis[start.y][start.x] = 1;
        
        while(!q.isEmpty()){
            Point p = q.poll();
            for(int i=0; i<4; i++){
                int nx = p.x + dx[i];
                int ny = p.y + dy[i];
                
                if(is_possible(nx, ny) && grid[ny][nx] == 1 && dis[ny][nx] == 0){
                    dis[ny][nx] = dis[p.y][p.x]+1;
                    q.add(new Point(nx, ny));
                }
            }
        }
        return dis[grid.length-1][grid[0].length-1];
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        grid = maps;
        
        // print(grid);
        answer = bfs(new Point(0, 0));
        if(answer == 0)
            answer = -1;
        return answer;
    }
}